doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +962,563 @@ tactic
+      .. coqtop:: reset all
 
-   .. tacv:: lazymatch goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
 
-      Using lazymatch instead of match will perform the same pattern matching
-      procedure but will commit to the first matching branch with the first
-      matching combination of hypotheses rather than trying a new matching if
-      the right-hand side fails. If the right-hand side of the selected branch
-      is a tactic with backtracking points, then subsequent failures cause
-      this tactic to backtrack.
+.. _match_goal_hyps_rev_ex:
+
+   .. example:: Matching hypotheses with reverse
+
+      Hypotheses are matched from the first hypothesis to the last until the :tacn:`apply` succeeds.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match reverse goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_multiple_hyps_ex:
+
+   .. example:: Multiple ways to match hypotheses
+
+      Every possible match for the hypotheses is evaluated until the right-hand
+      side succeeds.  Note that `H1` and `H2` are never matched to the same hypothesis.
+      Observe that the number of permutations can grow as the factorial
+      of the number of hypotheses and hypothesis patterns.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros A B H.
+         match goal with
+         | H1 : _, H2 : _ |- _ => idtac "match " H1 H2; fail
+         | _ => idtac
+         end.
+
+   .. todo need examples for:
+      match_context_rule ::= [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      match_hyp ::= | @name := {? [ @match_pattern ] : } @match_pattern
+
+.. todo PR The following items (up to numgoals) are part of "value_tactic".  I'd like to make
+   this a subsection and explain that they all return values.  How do I get a 5th-level section title?
 
 Filling a term context
 ~~~~~~~~~~~~~~~~~~~~~~
 
 The following expression is not a tactic in the sense that it does not
 produce subgoals but generates a term to be used in tactic expressions:
 
-.. tacn:: context @ident [@ltac_expr]
+.. tacn:: context @ident [ @term ]
+
+   .. todo @term is actually @lconstr (term with holes), which is spliced away
+      Seems like it's worth restoring that distinction
 
-   :n:`@ident` must denote a context variable bound by a context pattern of a
-   match expression. This expression evaluates replaces the hole of the
-   value of :n:`@ident` by the value of :n:`@ltac_expr`.
+   Replaces the context variable :n:`@ident` bound by a `context` pattern of a
+   match expression (Described :ref:`here<match_term_context>`).
+   This expression evaluates replaces the hole of the
```

Comment:
```suggestion
   This expression replaces the hole of the
```

Overall, this explanation is not clear and it seems like there are redundancies between the two sentences.